app/src/main/java/com/example/drawingactivity/GradientView.kt

```diff
@@ -0,0 +1,163 @@
+package com.example.drawingactivity
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.*
+import android.graphics.Color.BLACK
+import android.graphics.Color.HSVToColor
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import kotlinx.coroutines.*
+
+@SuppressLint("ClickableViewAccessibility")
+class GradientView @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
+) : View(context, attrs, defStyle) {
+
+    var onSatValChanged: ((Float, Float) -> Unit)? = null
+    var onGradientClicked: ((Boolean) -> Unit)? = null
+    private var hue = 360
+    private var shape: Rect? = null
+    private var dx = (width * 1f - 15f)
+    private var dy = 5f
+    private val selectPaint = Paint().apply {
+        strokeWidth = 3f
+        color = BLACK
+        style = Paint.Style.STROKE
+    }
+
+    init {
+        setOnTouchListener { _, mEvent ->
+            when (mEvent.action) {
+                MotionEvent.ACTION_DOWN -> {
+                    onGradientClicked?.invoke(true)
+                    dx = mEvent.x.coerceIn(0f, width * 1f)
+                    dy = mEvent.y.coerceIn(0f, height * 1f)
+                    if (dy > height / 2) {
+                        selectPaint.color = Color.WHITE
+                    } else selectPaint.color = BLACK
+                    onSatValChanged?.invoke(dx / width, dy / height)
+                    invalidate()
+                }
+                MotionEvent.ACTION_MOVE, MotionEvent.ACTION_CANCEL -> {
+                    dx = mEvent.x.coerceIn(0f, width * 1f)
+                    dy = mEvent.y.coerceIn(0f, height * 1f)
+                    if (dy > height / 2) {
+                        selectPaint.color = Color.WHITE
+                    } else selectPaint.color = BLACK
+                    onSatValChanged?.invoke(dx / width, dy / height)
+                    invalidate()
+                }
+                MotionEvent.ACTION_UP -> {
+                    onGradientClicked?.invoke(false)
+                }
+            }
+            true
+        }
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    private fun generateAllBitmaps() {
```

Comment:
This function should be in a companion object if it's setting variables in a (static) companion

Reply:
Moved this and the other function that impacts the constant into the companion object. also removed an unessicary null check